typing/typecore.ml

```diff
@@ -3814,11 +3814,15 @@ and type_function ?in_function loc attrs env mode ty_expected_explained l caseli
                                           ty_fun,
                                           explanation)))
   in
-  if not curry then
-    begin match Types.Alloc_mode.submode ret_mode Alloc_heap with
+  if curry then begin
+    match Types.Alloc_mode.submode mode ret_mode with
```

Comment:
This change is interesting. I guess it captures the idea that partially applying a function captures both the argument and the original function. This means that code like this isn't allowed:
```ocaml
let bar z =
  let[@stack]foo : int -> int -> int = fun x y -> x + y + z in
  ...
```
and it must instead be written as:
```ocaml
let bar z =
  let[@stack]foo : int -> (int -> int)[@stackret] = fun x y -> x + y + z in
  ...
```
I think this is another argument in favour of the proper currying support in types.

Reply:
Agreed!